File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void check(vector<int>&nums , vector<vector<int>>&v , vector<int>temp , int target , int idx ){
        if(target < 0 || idx >= nums.size()){
            return ; 
        } 
        if(target == 0){
            v.push_back(temp) ; 
            return ; 
        }
       
        temp.push_back(nums[idx]) ; 
        check(nums , v , temp , target-nums[idx] , idx) ;
        temp.pop_back() ; 
        check(nums , v , temp , target , idx+1) ;  
    }
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int>temp ; 
        vector<vector<int>>v; 

        check(nums , v , temp , target , 0) ; 
        return v ; 
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    void check(int[] nums , int idx , List<Integer>temp , int target , List<List<Integer>>v){
        if(idx>= nums.length || target < 0) {
            return ; 
        }
        if(target == 0){
             v.add(new ArrayList<>(temp)) ; 
             return ;
        }
        
        temp.add(nums[idx]) ; 
        check(nums , idx , temp , target - nums[idx], v) ; 
        temp.remove(temp.size()-1) ; 
        check(nums , idx+1 ,temp , target , v) ; 
       
    }
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>>v = new ArrayList<>() ; 
        List<Integer>temp = new ArrayList<>() ; 

        check(nums , 0 , temp , target  , v) ; 

        return v ; 
    }
}